orchagent/natorch.cpp

```diff
@@ -0,0 +1,4691 @@
+    snat_entry.data.mask.l4_src_port = 0xffff;
+    snat_entry.data.key.proto = ip_protocol;
+    snat_entry.data.mask.proto = 0xff;
+
+    status = sai_nat_api->remove_nat_entry(&snat_entry);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_INFO("Failed to removed %s SNAT NAPT entry with ip %s, port %d, prototype %s and it's translated ip %s, translated port %d",
+                      entry.entry_type.c_str(), keyEntry.ip_address.to_string().c_str(), keyEntry.l4_port, keyEntry.prototype.c_str(),
+                      entry.translated_ip.to_string().c_str(), entry.translated_l4_port);
+    }
+    else
+    {
+        SWSS_LOG_NOTICE("Removed %s SNAT NAPT entry with ip %s, port %d, prototype %s and it's translated ip %s, translated port %d",
+                      entry.entry_type.c_str(), keyEntry.ip_address.to_string().c_str(), keyEntry.l4_port, keyEntry.prototype.c_str(),
+                      entry.translated_ip.to_string().c_str(), entry.translated_l4_port);
+    }
+    deleteNaptCounters(keyEntry.prototype.c_str(), keyEntry.ip_address, keyEntry.l4_port);
+    m_naptEntries.erase(keyEntry);
+
+    if (entry.entry_type == "static")
+    {
+        if (totalStaticNaptEntries)
+        {
+            totalStaticNaptEntries--;
+            updateStaticNaptCounters(totalStaticNaptEntries);
+        }
+    }
+    else
+    {
+        if (totalDynamicNaptEntries)
+        {
+            totalDynamicNaptEntries--;
+            updateDynamicNaptCounters(totalDynamicNaptEntries);
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Found the total number dynamic napt entries to be corrupt, when removing SNAPT entry with proto %s, ip %s, port %d, translated ip %s, translated port %d!!",
+                            keyEntry.prototype.c_str(), keyEntry.ip_address.to_string().c_str(), keyEntry.l4_port,
+                            entry.translated_ip.to_string().c_str(), entry.translated_l4_port);
+        }
+    }
+
+    if (totalSnatEntries)
+    {
+        totalSnatEntries--;
+        updateSnatCounters(totalSnatEntries);
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Found the total number dynamic snat entries to be corrupt, when removing SNAPT entry with proto %s, ip %s, port %d, translated ip %s, translated port %d!!",
+                       keyEntry.prototype.c_str(), keyEntry.ip_address.to_string().c_str(), keyEntry.l4_port,
+                       entry.translated_ip.to_string().c_str(), entry.translated_l4_port);
+    }
+
+    if (totalEntries)
+        totalEntries--;
+
+    return true;
+}
+
+bool NatOrch::addNatEntry(const IpAddress &ip_address, const NatEntryValue &entry)
+{
+    SWSS_LOG_ENTER();
+
+    /* Check the entry is present in cache */
+    if (m_natEntries.find(ip_address) != m_natEntries.end())
+    {
+        SWSS_LOG_INFO("Duplicate %s %s NAT entry with ip %s and it's translated ip %s, do nothing",
+                      entry.entry_type.c_str(), entry.nat_type.c_str(), ip_address.to_stri
```

Comment:
In the current SONiC architecture, the kernel entries are updated/added/deleted from orchagent. Can you move this functionality to natmgr.

Reply:
OA is responsible for checking the hardware activity/hitbit of each of the entry in its cache. Since the cache is in OA code, it is the appropriate place to update the timeouts for the corresponding entries in the kernel.

Reply:
I understand OA is checking the hitbit, but in the current SONiC architecture the kernel entries are no update from OA. The mgr modules normally updates the kernel entries

Reply:
Agree with Arvind.

Reply:
The natmgr has no idea about the hardware activity of nat entries, so it cannot update the entries timeouts in kernel, if natmgr has to update the timeouts, OA has to inform the natmgr through separate communication channel which is an overhead and has it issues with scaling.

Reply:
The NAT entries are not updated real time, polling for hit-bit happens every 30 seconds, so adding another hop will not have a big impact on the performance.

Reply:
As discussed offline, lets keep the current code for now.  Let's open a issue to track this  change. We can make the code changes in the next release.